### ISSUE TITLE
msys2 shell error handling issues

### ISSUE THREAD

**Issue by lazka** [msys2/setup-msys2#43]
See the following workflow: https://github.com/lazka/msys2-workflow-error-test/actions/runs/167621530/workflow

The ubuntu bash runs by default with `set -e` (It's documented as `set -e o pipefail` here https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#exit-codes-and-error-action-preference but it doesn't look like that's true)

The msys2 shell doesn't which means a failing command doesn't fail the step, which is imo bad, at least for a default behavior in CI. It should at least default to `-e`

Any ideas?

**Comment by lazka** [comment]
> (It's documented as `set -e o pipefail` here https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#exit-codes-and-error-action-preference but it doesn't look like that's true)

This was already reported here https://github.com/actions/runner/issues/353

**Comment by eine** [comment]
I'm not sure I understand the context... Your concern is legit, and I agree. But, please, let me elaborate:

- Commands executed through this Action's `msys2 {0}` are called with `bash --norc -ilceo pipefail` (see https://github.com/msys2/setup-msys2/blob/master/main.js#L74). Hence, regardless of the default setup of the shells provided by GitHub, the behaviour of the Action should be as expected.
- If users `set -e` in their scripts, the behaviour is the expected.

I wrote multiple tests in a fork of your demo (https://github.com/eine/msys2-workflow-error-test/blob/master/.github/workflows/main.yml), but I reached no conclusive solution:

- `shell: bash` is converted to `shell: C:\Program Files\Git\bin\bash.EXE --noprofile --norc -e -o pipefail {0}` and it works as expected.
  - However, writing it manually does not work: https://github.com/eine/msys2-workflow-error-test/runs/866200010?check_suite_focus=true#step:3:4
- `bash --norc -ilceo pipefail {0}` fails because of the slashes: https://github.com/eine/msys2-workflow-error-test/runs/866200222?check_suite_focus=true#step:3:7

Might this issue be related to this warning https://github.com/eine/msys2-workflow-error-test/runs/866200165?check_suite_focus=true#step:4:9 ?

**Comment by eine** [comment]
> This was already reported here actions/runner#353

I think that's a similar but unrelated issues. It is about getting a different behaviour when not setting `shell` or when setting it to `bash`. In our case, we don't use any of them, neither the default nor bash. Our entrypoint is `msys2.cmd`.

**Comment by lazka** [comment]
> > This was already reported here [actions/runner#353](https://github.com/actions/runner/issues/353)
> 
> I think that's a similar but unrelated issue. It is about getting a different behaviour when not setting `shell` or when setting it to `bash`. In our case, we don't use any of them, neither the default nor bash. Our entrypoint is `msys2.cmd`.

Yeah, unrelated, I just wanted to suggest at first that we should mirror what bash does in GHA, but turned out that it wasn't that clear...

**Comment by lazka** [comment]
>     * Commands executed through this Action's `msys2 {0}` are called with `bash --norc -ilceo pipefail` (see https://github.com/msys2/setup-msys2/blob/master/main.js#L74). Hence, regardless of the default setup of the shells provided by GitHub, the behaviour of the Action should be as expected.

Doesn't it pass a file there? That would mean a new shell.

**Comment by eine** [comment]
My understanding is that `run: anything` is passed raw. That's why `run: anything && anyotherthing` and such might fail. However, the content of `run: |` is saved to some temporal location and passed as a file. See numworks/setup-msys2#22.

Might be worth trying `run: fail; echo "ok";`?

**Comment by eine** [comment]
> Doesn't it pass a file there? That would mean a new shell.

Oh, I see what you mean. If `run: |`, https://github.com/msys2/setup-msys2/blob/master/main.js#L74 might need to be:

```
drive + `\\msys64\\usr\\bin\\env.exe /usr/bin/bash --norc -ilceo pipefail "cd $OLDPWD && bash -e %args%"`
```

or something similar.

**Comment by lazka** [comment]
I think I have something, I'll create a PR later

**Comment by lazka** [comment]
oh, meh, it gets used to run scripts and binary files (like to "go env" example). No idea how to fix that without breaking things.

**Comment by eine** [comment]
@laxka, I guess this is fixed, now that #44 is merged, isn't it?

**Comment by lazka** [comment]
yes!
